velox/exec/tests/UnnestTest.cpp

```diff
@@ -28,15 +28,59 @@ TEST_F(UnnestTest, basicArray) {
           100,
           [](auto row) { return row % 5 + 1; },
           [](auto row, auto index) { return index * (row % 3); },
-          nullEvery(7)),
+          nullEvery(10)),
   });
 
   createDuckDbTable({vector});
 
   // TODO Add tests with empty arrays. This requires better support in DuckDB.
 
   auto op = PlanBuilder().values({vector}).unnest({"c0"}, {"c1"}).planNode();
-  assertQuery(op, "SELECT c0, UNNEST(c1) FROM tmp WHERE c0 % 7 > 0");
+  assertQuery(op, "SELECT c0, UNNEST(c1) FROM tmp WHERE c0 % 10 > 0");
+
+  auto opWithOrdinality =
+      PlanBuilder().values({vector}).unnest({"c0"}, {"c1"}, "c1").planNode();
+
+  // From a row number in the result column, calculate the original row number
+  // and index of an element in the original array column being unnested.
+  auto getOriginalRowAndIndex = [](auto row) {
+    // Calculate how many 5 rows in the original array vector are there before
+    // the original row that contains the element at this result row. The
+    // calculation is based on the following facts: there are 15 elements in
+    // every 5 rows in the original array vector; null rows in the original
+    // vector translate to nothing in the unnest result; those null rows should
+    // contain 1 element if they were not null.
+    int originalRow = (row + (row / 29 + 1)) / 15 * 5;
+    int originalIndex = 0;
+
+    // Calculate the exact row number and index of the element inside the
+    // original array accoding to the offset of this element in the result.
+    row = (row + (row / 29 + 1)) % 15;
+    int offset[5] = {0, 1, 3, 6, 10};
+    for (int i = 4; i >= 0; --i) {
+      if (row >= offset[i]) {
+        originalRow += i;
+        originalIndex = row - offset[i];
+        break;
+      }
+    }
+    return std::make_pair(originalRow, originalIndex);
+  };
+
+  auto expected = makeRowVector(
+      {makeFlatVector<int64_t>(
+           290, [&](auto row) { return getOriginalRowAndIndex(row).first; }),
+       makeFlatVector<int32_t>(
+           290,
+           [&](auto row) {
+             auto originalRowAndIndex = getOriginalRowAndIndex(row);
+             return originalRowAndIndex.second *
+                 (originalRowAndIndex.first % 3);
+           }),
+       makeFlatVector<int64_t>(290, [&](auto row) {
+         return getOriginalRowAndIndex(row).second + 1;
+       })});
+  assertQuery(opWithOrdinality, expected);
```

Comment:
Why are we not using a duckDB query here (instead of comparing with a constructed result vector) ?

Reply:
Thank you for the question. I didn't use duckDB because it doesn't support "with ordinality" in the SQL query. (I tried and got an error.)